onnxruntime/core/providers/webnn/webnn_execution_provider.cc

```diff
@@ -10,13 +10,101 @@
 #include "core/graph/graph_viewer.h"
 #include "core/session/onnxruntime_cxx_api.h"
 #include "core/common/safeint.h"
+#include "core/common/inlined_containers.h"
+#include "core/common/hash_combine.h"
 
 #include "builders/model.h"
 #include "builders/helper.h"
 #include "builders/model_builder.h"
 
 namespace onnxruntime {
 
+struct WebNNContextOptions {
+  std::optional<std::string> device_type;
+  std::optional<int> num_threads;
+  std::optional<std::string> power_preference;
+
+  [[nodiscard]] emscripten::val AsVal() const {
+    emscripten::val options = emscripten::val::object();
+    if (device_type.has_value()) {
+      options.set("deviceType", device_type.value());
+    }
+    if (num_threads.has_value()) {
+      options.set("numThreads", num_threads.value());
+    }
+    if (power_preference.has_value()) {
+      options.set("powerPreference", power_preference.value());
+    }
+    return options;
+  }
+
+  bool operator==(const WebNNContextOptions& other) const {
+    return std::tie(device_type, num_threads, power_preference) ==
+           std::tie(other.device_type, other.num_threads, other.power_preference);
+  }
+};
+
+}  // namespace onnxruntime
+
+// Specialize std::hash for WebNNContextOptions.
+template <>
+struct std::hash<::onnxruntime::WebNNContextOptions> {
+  size_t operator()(const ::onnxruntime::WebNNContextOptions& options) const {
+    size_t hash{0xbc9f1d34};  // seed
+    if (options.device_type.has_value()) {
+      onnxruntime::HashCombine(options.device_type.value(), hash);
+    }
+    if (options.num_threads.has_value()) {
+      onnxruntime::HashCombine(options.num_threads.value(), hash);
+    }
+    if (options.power_preference.has_value()) {
+      onnxruntime::HashCombine(options.power_preference.value(), hash);
+    }
+    return hash;
+  }
+};
+
+namespace onnxruntime {
+
+// WebNNContextManager is a singleton object that is used to create MLContexts.
+class WebNNContextManager {
+ public:
+  static WebNNContextManager& GetInstance() {
+    static WebNNContextManager instance;
+    return instance;
+  }
+
+  WebNNContextManager(const WebNNContextManager&) = delete;
+  WebNNContextManager& operator=(const WebNNContextManager&) = delete;
+  WebNNContextManager(WebNNContextManager&&) = delete;
+  WebNNContextManager& operator=(WebNNContextManager&&) = delete;
+
+  emscripten::val GetContext(const WebNNContextOptions& options) {
+    auto it = contexts_.find(options);
+    if (it != contexts_.end()) {
+      return it->second;
+    }
+
+    emscripten::val ml = emscripten::val::global("navigator")["ml"];
+    if (!ml.as<bool>()) {
+      ORT_THROW("Failed to get ml from navigator.");
+    }
+
+    emscripten::val context = ml.call<emscripten::val>("createContext", options.AsVal()).await();
```

Comment:
So how do you get global shared context from JS to create ml buffer?

Reply:
That's a good point, this PR does not expose the context to JS. We'll need access to MLContext from JS to enable ort-web to download and upload data to the MLBuffer from JS.

@fs-eire do you have any suggestion on how to solve this issue without adding a WebNN TS backend?

Reply:
I am not familiar with the Embind API, but I think the idea is the same:

Using the [EMSCRIPTEN_BINDINGS()](https://emscripten.org/docs/api_reference/bind.h.html#_CPPv419EMSCRIPTEN_BINDINGS4name) macro to export C++ functions to the [`Module object`](https://emscripten.org/docs/api_reference/module.html). For example, a function called `getCurrentMLContext` (takes no parameter) or `getMLContext` (takes session ID as parameter) and returns the reference of the object. Then, you can call `Module['getCurrentMLContext']()` in js_internal_api.js (because this file is included in the final JS glue using `--pre-js` in emcc.

Please see also:
- https://emscripten.org/docs/api_reference/bind.h.html#_CPPv419EMSCRIPTEN_BINDINGS4name
- https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html#a-quick-example
-

Reply:
Considering the actual usage of IO binding, users may need to access the MLContext object via ORT javascript API. It can be something like:
```ts

const mySession = await ort.InferenceSession.create('...', { executionProviders: [{
  name: 'webnn',
  ...
}] });

const myMLContext = ort.env.webnn.getContext(mySession);

```

Considering this, there need to be a [SessionObject(JS)] to [MLContext(JS)] mapping in JavaScript. We can get [SessionObject(JS)] to [SessionID(JS&C++)] mapping, and [SessionID(JS&C++)] to [MLContext(JS)] mapping, so this is able to implement.

Reply:
Unless I'm missing something, there is no public API to get the `WebNNExecutionProvider*`  from the [SessionID(JS&C++)](i.e. `OrtSession*`) and `Module.jsepSessionState.sessionHandle` is only valid during `run()`.

Reply:
I see your point. After checking the code, I realize that using session ID is not a good idea. There is no existing places that we can associate a session handle with a Web NN EP instance in the code.

In C++ we can just expose the `WebNNContextManager::GetContext()` to JavaScript.

And in JavaScript API, we have 2 options:

- Let user pass options directly:
  ```ts
  const myWebNNOptions = {
    name: 'webnn',
    ...
  };
  const mySession = await ort.InferenceSession.create('...', {
    executionProviders: [myWebNNOptions ]
  });

  const myMLContext = ort.env.webnn.getContext(myWebNNOptions);
  ```
  - No need to add extra JS code

- Let user pass session object:
  ```ts
  const mySession = await ort.InferenceSession.create('...', { executionProviders: [{
    name: 'webnn',
    ...
  }] });

  const myMLContext = ort.env.webnn.getContext(mySession);
  ```
  - May be little bit easier to use for users, but require maintain a session-to-webnnOptions map in JS

A few changes to init.ts, backend-webnn.ts may have to add back, as we need to allow to assign object `ort.env.webnn` to a property of `Module` so that in C++ embind can access the object and add a property ( the `getContext()` function) to it.